pkg/knative/logs.go

```diff
@@ -24,126 +17,6 @@ import (
-	if err != nil {
-		return fmt.Errorf("cannot create k8s client: %w", err)
-	}
-
-	pods := client.CoreV1().Pods(namespace)
-
-	podListOpts := metav1.ListOptions{
-		Watch:         true,
-		LabelSelector: fmt.Sprintf("serving.knative.dev/service=%s", kServiceName),
-	}
-
-	w, err := pods.Watch(ctx, podListOpts)
-	if err != nil {
-		return fmt.Errorf("cannot create watch: %w", err)
-	}
-	defer w.Stop()
-
-	beingProcessed := make(map[string]bool)
-	var beingProcessedMu sync.Mutex
-
-	copyLogs := func(pod corev1.Pod) error {
-		defer func() {
-			beingProcessedMu.Lock()
-			delete(beingProcessed, pod.Name)
-			beingProcessedMu.Unlock()
-		}()
-		podLogOpts := corev1.PodLogOptions{
-			Container: "user-container",
-			Follow:    true,
-		}
-		if since != nil {
-			sinceTime := metav1.NewTime(*since)
-			podLogOpts.SinceTime = &sinceTime
-		}
-		req := client.CoreV1().Pods(namespace).GetLogs(pod.Name, &podLogOpts)
-
-		r, e := req.Stream(ctx)
-		if e != nil {
-			return fmt.Errorf("cannot get stream: %w", e)
-		}
-		defer r.Close()
-		_, e = io.Copy(out, r)
-		if e != nil {
-			return fmt.Errorf("error copying logs: %w", e)
-		}
-		return nil
-	}
-
-	mayReadLogs := func(pod corev1.Pod) bool {
-		for _, status := range pod.Status.ContainerStatuses {
-			if status.Name == "user-container" {
-				return status.State.Running != nil || status.State.Terminated != nil
-			}
-		}
-		return false
-	}
-
-	getImage := func(pod corev1.Pod) string {
-		for _, ctr := range pod.Spec.Containers {
-			if ctr.Name == "user-container" {
-				return ctr.Image
-			}
-		}
-		return ""
-	}
-
-	var eg errgroup.Group
-
-	for event := range w.ResultChan() {
-		if event.Type == watch.Modified || event.Type == watch.Added {
-			pod := *event.Object.(*corev1.Pod)
-
-			beingProcessedMu.Lock()
-			_, loggingAlready := beingProcessed[pod.Name]
-			beingProcessedMu.Unlock()
-
-			if !loggingAlready && (image == "" || image == getImage(pod)) && mayReadLogs(pod) {
-
-				beingProcessedMu.Lock()
-				beingProcessed[pod.Name] = true
-				beingProcessedMu.Unlock()
-
-				eg.Go(func() error { return copyLogs(pod) })
-			}
-		}
-	}
-
-	err = eg.Wait()
-	if err != nil {
-		return fmt.Errorf("error while gathering logs: %w", err)
-	}
-	return nil
-}
-
-type SynchronizedBuffer struct {
-	b  bytes.Buffer
-	mu sync.Mutex
-}
-
-func (b *SynchronizedBuffer) String() string {
-	b.mu.Lock()
-	defer b.mu.Unlock()
-	return b.b.String()
-}
-
-func (b *SynchronizedBuffer) Write(p []byte) (n int, err error) {
-	b.mu.Lock()
-	defer b.mu.Unlock()
-	return b.b.Write(p)
-}
-
-func (b *SynchronizedBuffer) Read(p []byte) (n int, err error) {
-	b.mu.Lock()
-	defer b.mu.Unlock()
-	return b.b.Read(p)
-}
-
-func (b *SynchronizedBuffer) Reset() {
-	b.mu.Lock()
-	defer b.mu.Unlock()
-	b.b.Reset()
+	selector := fmt.Sprintf("serving.knative.dev/service=%s", kServiceName)
+	return k8s.GetPodLogsBySelector(ctx, namespace, selector, "user-container", image, since, out)
```

Comment:
Nice that this is also now shared; we can use it as a starting point for proper logging facilities through the client (a long-standing feature request!)